### PR TITLE
Semantic snippets - `propg` and `propi` snippets

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Abstracted a lot of logic for property generation. In theory, it now allows to generate any kind of auto property, even with different `get` accessor.

On top of new abstractions added 2 snippets:
- `propg` is a replacement of an old snippet. It generates property like `public int MyProperty { get; private set; }`. The only thing I changed is that when this snippet is executed in interface, it doesn't generate `private set` accessor, because it leads to compiler error. If a user executes `propg` snippet in interface I am 99% sure he wants just a `get` accessor
- `propi` is a new snippet I invented "on the fly". Usage of `init` properties has grown over years and after `required` modifier from C# 11 I assume it would grow even more. I don't see valid reasons against adding `init`-only property snippet then

Abstracted property snippet tests as well. There are now 11 basic cases and an ability to override any special case behaviour like it is done for `propg` inside an interface.

I got a bit carried away, so unfortunately the first commit is bit large and could be 100% splitted into smaller parts, sorry for that!

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
Can I get a review here? @akhera99 ?

**Comment by akhera99** [comment]
> Can I get a review here? @akhera99 ?

Looking!

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpAutoPropertyCompletionProviderTests.cs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            await VerifyDefaultPropertyAsync("""
+                struct MyStruct
+                {
+                    $$
+                }
+                """);
+        }
+
+        [WpfFact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public virtual async Task InsertSnippetInInterface()
```
Comment (akhera99) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpAutoPropertyCompletionProviderTests.cs:79]: I think I would prefer just writing this separately for the different prop snippets

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpAutoPropertyCompletionProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                {
+                    public Program()
+                    {
+                        $$
+                    }
+                }
+                """);
+        }
+
+        protected abstract string GetDefaultPropertyText(string propertyName);
```
Comment (akhera99) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/Snippets/CSharpAutoPropertyCompletionProviderTests.cs:140]: can you move this to the top of the file

**Review by akhera99 — APPROVED** [review]
Looks good to me other than the things noted!

**Comment by akhera99** [comment]
Merged, thanks for the contribution!

**Pull request merged**